plutus-core/plutus-core/src/Data/SatInt.hs

```diff
@@ -246,21 +265,52 @@ quotRemSI a@(I# _) b@(I# _) = (SI (a `quotInt` b), SI (a `remInt` b))
 divModSI ::  Int -> Int -> (SatInt, SatInt)
 divModSI x@(I# _) y@(I# _) = (SI (x `divInt` y), SI (x `modInt` y))
 
+
 plusSI :: SatInt -> SatInt -> SatInt
+plusSI (SI (I# x#)) (SI (I# y#)) =
+  case addIntC# x# y#  of
+    (# r#, f# #) ->
+        if isTrue# f# then  -- Overflow;  I think the value of f# is always 1 in this case
+                            -- and doesn't give us any other useful information
+        if      isTrue# ((x# ># 0#) `andI#` (y# ># 0#)) then maxBound
+        else if isTrue# ((x# <# 0#) `andI#` (y# <# 0#)) then minBound
+        else 0  -- Shouldn't happen: error?  `succ maxBound` already raises an exception.
+        else SI (I# r#)
+
+{- **** I originally tried
 plusSI (SI (I# x#)) (SI (I# y#)) =
   case addIntC# x# y# of
-    (# r#, _ #) -> SI (I# r#)
+    (# r#, 0# #) -> SI (I# r#)
+    _ -> if isTrue# (x# ># 0#) && isTrue# (y# ># 0#) then maxBound
+         else ...
+
+but that made the benchmarks slower, maybe because it has to go via Eq.
+
+I believe that things like ># return either 0 or 1, so it's safe to use bitwise
+`and#` here rather than `isTrue# (x# ># 0#) && isTrue# (y# ># 0#)`.  That only
+matters when we've overflowed though, so it's probably not too important to be
+super-efficient in that case.
+ **** -}
 
 minusSI :: SatInt -> SatInt -> SatInt
 minusSI (SI (I# x#)) (SI (I# y#)) =
-  case subIntC# x# y# of
-    (# r#, _ #) -> SI (I# r#)
+    case subIntC# x# y# of
+      (# r#, f# #) ->
+          if isTrue# f# then -- Overflow
+          if      isTrue# ((x# >=# 0#) `andI#` (y# <# 0#)) then maxBound
```

Comment:
> Any reason for the `>=` here versus `>` above?

I think that for `plusSi` we could equally well put `>=`, but that would only apply when one of the arguments is 0, and surely you can't get an overflow if you're adding 0 to something (although I'm beginning to have my doubts about that now).

We do need `>=` here because `0-minBound` should overflow (`minBound` is -2^63, `maxBound` is 2^63-1).  I think it'd be safe to change `<#`  into `<=#` here though.

Note that there's also a `negateInt#` operation, and the [documentation](https://hackage.haskell.org/package/ghc-prim-0.7.0/docs/GHC-Prim.html) for that says

> negateInt# of the most negative number is an identity operation. This way, negateInt# is always its own inverse.

In other words, `-minBound = minBound` (and both are negative), which I think also happens in Java, C, and so on.  This doesn't happen with the definitions in this file.  The `negate` instance gives `-minBound = maxBound`  and the code above gives `0-minBound = maxBound`.